### DISCUSSION TITLE
Do you have an event dataset you can share

### DISCUSSION THREAD

**Discussion by veonua** [dadaloop82/MyHomeSmart-HASS-AppDeamon#8] [General]
Data is a key for ML project, if you can share one people can try their hypothesis and come up with ideas

**Comment by dadaloop82** [comment]
Hi @veonua 

You're absolutely right, data is the essential key to making an ML work.

In my code the data is already collected, in real time, in a sqlite3 db and is stored in an optimized form.
In the DB "entitystate_DB" for example we have 3 tables, one collects the entities, one the states and the last one is the union between these two with the date of the event.

The storage of the states is optimized so that if there are states in numeric form it doesn't store the single values but the range of them and if a single value repeats itself a smaller group with that value is created.
I believe that this system, besides being intelligent, can also be useful to determine the user's preferred value.

These days I'm focusing on the determination of cause and effect, that is the creation of the chain of events that led to a certain effect.
For example, if I open the window and it is cold outside, the consequence is that the temperature in the house goes down.
This could serve me to determine the best choice in order to catch up a determined objective or to estimate if the previewed action could be useful for the objective.

I am not an expert in ML, in fact I know very little about it, your help could be really important for the development of this project.
For example at the moment I don't know how to interface this information with SciKit - but I'm sure you will give me the right advice!

Thank you for your time!
Daniel

**Reply by veonua** [reply]
> The storage of the states is optimized so that if there are states in numeric form it doesn't store the single values but the range of them and if a single value repeats itself a smaller group with that value is created.
I believe that this system, besides being intelligent, can also be useful to determine the user's preferred value.

premature optimization ... you know

> For example, if I open the window and it is cold outside, the consequence is that the temperature in the house goes down.

can you tell this in terms of the data, which sensors, and measurements you get? What is the feedback, and how do you measure the effect?

**Reply by dadaloop82** [reply]
> premature optimization ... you know

I know, it could be something that plays against me, but stupid to store states like this in raw format: in this way many states maybe used only once would consume space in DB for nothing and I don't think it's practical to processing.
I don't know if I'll keep it, for now it seems like a good idea, but of course I'm available for any advice.

> can you tell this in terms of the data, which sensors, and measurements you get? What is the feedback, and how do you measure the effect?

I don't have any data in hand at the moment, it's all in the "thinking" and testing phase.
My idea is to create a chain of events, as reported in a very simplistic form in my chart.
In a few words the sequence should be this

example
```
(T1) : outdoor temperature sensor
(T2) : internal temperature sensor
(S1) : heating switch (initialy OFF)
(S2) : air conditioner switch (initialy OFF)
```

Data:

| Events| #1  | #2  | #3  | #4  | #5  | #6  | #7  | #8  |
| :---:   | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| T1| 20 |  | 19 | 17 | 19 | | 22 |
| T2|  | 21 | 20 | | 17 | | | 21 |
| S1|  | | | | ON | | | OFF |
| S2|  | ON | | | OFF | | ON | |


Chains:

#1 (from event #1) T1 -> (from event #2) T2 -> (from event #2) S2 
#2 (from event #3) T1 -> (from event #3) T2 
#3 (from event #5) T1 -> (from event #5) T2  -> (from event #5) S1  -> (from event #5) S2
#4 (from event #7) T1 -> (from event #7) S2
#5 (from event #8) T2 -> (from event #7) S1

Consider that in the above (entirely simplistic) example, the "inverse chains" are not shown, i.e. the effect of when you turn off the air conditioner or the house heater.
I believe that in this way I can determine the cause->effect link of the various state changes, which is essential to define a scenario of awareness of actions to be taken.

What do you think?

**Reply by veonua** [reply]
ok, we have these observations. What is the ML algorithm supposed to do and when?


form the table it's not really clear if these are events... why there are two events at 5 and zero events at 6.
But I would recommend thinking in terms of states, let's say every 5 seconds ML algorithm takes a full snapshot of the system state and state duration as the input. 
My idea is to predict the next state based on the previous state (or many states).

**Reply by dadaloop82** [reply]
Good morning!
So many great ideas and suggestions! Thank you!

So: in the table I made you, in a totally simplistic form, there are empty cells because I currently take the data in live not via snapshot as you suggested. Your idea to make deli snapshot, that is to have the current situation, is very interesting and could be combined with my idea. 
I will put it in the ToDo list!

It seems to me that you are very knowledgeable about ML, and this could help me a lot.
But I have a doubt that an ML doesn't need just raw snapshot's data - I don't understand how it could understand the link between entities and its own state changes.

Taking the most trivial example, how would it know that turning on the heat would raise the temperature of the house and therefore the windows would have to be closed (if they are smart?).
In my head I would have in mind to give this information to the ML (more precisely to a Decision-Tree) to make the most optimized choice to reach the goal.

What do you think?

**Reply by veonua** [reply]
there are several approaches to this problem, 
1. programmatic - you bring knowledge to the system
>  if (group.temperature.drops && group.temperature < 17) then group.heater.turn_on()
>    if (....)
  takes many hours of human, and it's explainable at some level

2. local ML model - stream events to the system and let it find patterns. It takes a lot of user-specific data, if the user turns on the heater regularly when the temperature drops under 17* the system will spot this pattern eventually. it will take some time and data for the system to learn basic things "temperature rising" "heater is the device for heating"

3. global ML model - stream events from many users. The amount of data is more by definition, so find rare patterns likely already there. Patterns among humans are likely to be the same. Lightbulbs, windows and heaters do the same thing around the globe.

the real system is a mix of these strategies.

**Reply by dadaloop82** [reply]
Wow! Very interesting, thank you!
You've opened up a whole new perspective for me!


It's very interesting your explanation about local and global ML system. I would absolutely say to implement your idea (hopefully with your help) and define the flow in this generalized way:

1) data collection
2) saving and optimizing the collected data
3) sending data to the Local ML model
3a) with the time the system will identify which are the conditions for which a certain entity varies its state
3b) with the time the system will learn also from the habits of the customer, that is that as an example to 8 of morning it turns on the light of the room.
4) shipment of the data (anonymous) to a api web in order to find rare or predefined models (like you say, the bulbs or other make the same thing and they have need of similar conditions in order to change of state
4a) processing of global ML model on remote machine
5) identification of current local situation with reasoning on actions to take
6) if the action has not enough weight, ask the user (via alexa?) if you want that action to be executed (ML Local)
7) execute the action

What do you think? 
How would you change it?

I would love to collaborate with you to achieve this common goal.
Forgive me if in ML topic I am not very practical, but I have good ideas :)

Thanks!

**Reply by veonua** [reply]
Yes, overall the flow looks right.
 
I prefer cloud implementation first. just because you can use a much wider and better toolset and not care about performance and resources. Also, I'd recommend postponing optimization until you have something working.


I don't have Alexa and I'm not sure if it supports proactive notifications. I've been trying to see if my Google Nest can do this and can't see an easy way. So I thought more about web notifications or mobile notifications. This is something that requires research.


In the beginning, my goal is to collect low hanging fruits and provide useful features:
Answer simple questions based on the data - what happened while I was sleeping, while I was away.
Anomaly detection - forgotten lights, doors, windows. Movement when/where it shouldn't be.
